src/portable/espressif/esp32s2/dcd_esp32s2.c

```diff
@@ -0,0 +1,749 @@
+    // DOEPINT will be cleared when DAINT's out bits are cleared.
+    for (int n = 0; n < USB_OUT_EP_NUM; n++) {
+        xfer_ctl_t *xfer = XFER_CTL_BASE(n, TUSB_DIR_OUT);
+        if (USB0.daint & (1 << (16 + n))) {
+            // SETUP packet Setup Phase done.
+            if ((USB0.out_ep_reg[n].doepint & USB_SETUP0_M)) {
+                USB0.out_ep_reg[n].doepint |= USB_STUPPKTRCVD0_M | USB_SETUP0_M; // clear
+                if (s_setup_phase == 1) {                                          // only if setup is done, but not handled
+                    s_setup_phase = 2;
+                    ESP_EARLY_LOGV(TAG, "TUSB IRQ - setup_phase 2"); // sending to  a handling queue
+                    ESP_EARLY_LOGV(TAG, "TUSB IRQ - EP OUT - Setup Phase done (irq-s 0x%08x)", USB0.out_ep_reg[n].doepint);
+                    dcd_event_setup_received(0, (uint8_t *)&_setup_packet[0], true);
+                }
+                readyfor1setup_pkg(0);
+            }
+
+            // OUT XFER complete (single packet).q
+            if (USB0.out_ep_reg[n].doepint & USB_XFERCOMPL0_M) {
+
+                ESP_EARLY_LOGV(TAG, "TUSB IRQ - EP OUT - XFER complete (single packet)");
+                USB0.out_ep_reg[n].doepint = USB_XFERCOMPL0_M;
+
+                // Transfer complete if short packet or total len is transferred
+                if (xfer->short_packet || (xfer->queued_len == xfer->total_len)) {
+                    xfer->short_packet = false;
+                    dcd_event_xfer_complete(0, n, xfer->queued_len, XFER_RESULT_SUCCESS, true);
+                } else {
+                    // Schedule another packet to be received.
+                    USB0.out_ep_reg[n].doeptsiz = USB_PKTCNT0_M |
+                                                  ((xfer->max_size & USB_XFERSIZE0_V) << USB_XFERSIZE0_S);
+                    USB0.out_ep_reg[n].doepctl |= USB_EPENA0_M | USB_CNAK0_M;
+                }
+            }
+        }
+    }
+}
+
+static void handle_epin_ints(void)
+{
+
+    // GINTSTS will be cleared with DAINT == 0
+    // DAINT for a given EP clears when DIEPINTx is cleared.
+    // IEPINT will be cleared when DAINT's out bits are cleared.
+    for (uint32_t n = 0; n < USB_IN_EP_NUM; n++) {
+        xfer_ctl_t *xfer = &xfer_status[n][TUSB_DIR_IN];
+
+        if (USB0.daint & (1 << (0 + n))) {
+            ESP_EARLY_LOGV(TAG, "TUSB IRQ - EP IN %u", n);
+            // IN XFER complete (entire xfer).
+            if (USB0.in_ep_reg[n].diepint & USB_D_XFERCOMPL0_M) {
+                ESP_EARLY_LOGV(TAG, "TUSB IRQ - IN XFER complete!");
+                USB0.in_ep_reg[n].diepint = USB_D_XFERCOMPL0_M;
+                // USB0.dtknqr4_fifoemptymsk &= ~(1 << n); // Turn off TXFE b/c xfer inactive.
+                dcd_event_xfer_complete(0, n | TUSB_DIR_IN_MASK, xfer->total_len, XFER_RESULT_SUCCESS, true);
+            }
+
+            // XFER FIFO empty
+            if (USB0.in_ep_reg[n].diepint & USB_D_XFERCOMPL0_M) {
```

Comment:
@cr1901 spotted anything suspicious :thinking: :thinking:  It took me a day with tons of trials and reading specs back and forth to catch this copy-paste typo. Man, that is so typical, phew doing some clean up. there is still a bit of issue with other thing. But It is quite a progress for today :)

Reply:
Copy and paste typos happen. Good that you caught it though- because I certainly didn't :P.

Reply:
phew, I feel so lucky catching this typo :)

Reply:
funny thing, correct the typo make thing worse, esp32s2 doesn't pass enumeration now. It now only responds with exactly 1 packet. I better to rest now, probably made some typo/mistake in the progress :sweat_smile: :sweat_smile: